azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -73,44 +67,58 @@ def _execute_op(self, op):
             self.transport.on_mqtt_connection_failure_handler = self._on_mqtt_connection_failure
             self.transport.on_mqtt_disconnected_handler = self._on_mqtt_disconnected
             self.transport.on_mqtt_message_received_handler = self._on_mqtt_message_received
-            self._active_connect_op = None
-            self._active_disconnect_op = None
+
+            # There can only be one pending connection operation (Connect, Reconnect, Disconnect)
+            # at a time. The existing one must be completed or canceled before a new one is set.
+
+            # Currently, this means that if, say, a connect operation is the pending op and is executed
+            # but another connection op is begins by the time the CONACK is received, the original
+            # operation will be cancelled, but the CONACK for it will still be received, and complete the
+            # NEW operation. This is not desirable, but it is how things currently work.
+
+            # We are however, checking the type, so the CONACK from a cancelled Connect, cannot successfully
+            # complete a Disconnect operation.
+            self._pending_connection_op = None
+
             self.pipeline_root.transport = self.transport
```

Comment:
>self.pipeline_root.transport = self.transport [](start = 12, length = 45)

By the way, why does this happen? I'm not clear on the purpose.

Reply:
removed in my PR

---
In reply to: [311703157](https://github.com/Azure/azure-iot-sdk-python-preview/pull/159#discussion_r311703157) [](ancestors = 311703157)